test/fixtures/azure-fixture/src/main/java/fixture/azure/MockAzureBlobStore.java

```diff
@@ -0,0 +1,487 @@
+        }
+    }
+
+    private static class MockAzureBlockBlob extends AbstractAzureBlob implements AzureBlob {
+        private final Map<String, BytesReference> blocks;
+        private BytesReference contents = EMPTY;
+
+        private MockAzureBlockBlob() {
+            this.blocks = new ConcurrentHashMap<>();
+        }
+
+        public void putBlock(String blockId, BytesReference content, String leaseId) {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                this.blocks.put(blockId, content);
+            }
+        }
+
+        public void putBlockList(List<String> blockIds, String leaseId) throws BadRequestException {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                final List<String> unresolvedBlocks = blockIds.stream().filter(bId -> blocks.containsKey(bId) == false).toList();
+                if (unresolvedBlocks.isEmpty() == false) {
+                    logger.warn("Block list contained non-existent block IDs: {}", unresolvedBlocks);
+                    throw new BadRequestException("InvalidBlockList", "The specified blocklist is invalid.");
+                }
+                final BytesReference[] resolvedContents = blockIds.stream().map(blocks::get).toList().toArray(new BytesReference[0]);
+                contents = CompositeBytesReference.of(resolvedContents);
+            }
+        }
+
+        private boolean matches(String ifNoneMatchHeaderValue) {
+            if (ifNoneMatchHeaderValue == null) {
+                return false;
+            }
+            // We only support *
+            if ("*".equals(ifNoneMatchHeaderValue)) {
+                return true;
+            }
+            throw new BadRequestException("UnsupportedHeader", "The test fixture only supports * for if-none-match");
+        }
+
+        @Override
+        public synchronized void setContents(BytesReference contents, String leaseId) {
+            synchronized (writeLock) {
+                lease.assertCanWrite(leaseId);
+                this.contents = contents;
+                this.blocks.clear();
+            }
+        }
+
+        @Override
+        public void setContents(BytesReference contents, String leaseId, String ifNoneMatchHeaderValue) {
+            synchronized (writeLock) {
+                if (matches(ifNoneMatchHeaderValue)) {
+                    throw new PreconditionFailedException(
+                        "TargetConditionNotMet",
+                        "The target condition specified using HTTP conditional header(s) is not met."
+                    );
+                }
+                setContents(contents, leaseId);
+            }
+        }
+
+        @Override
+        public BytesReference getContents() {
+            return contents;
+        }
+
+        @Override
+        public long length() {
+            return contents == null ? 0 : contents.length();
+        }
```

Comment:
Do we need synchronization here? It reads `contents` twice and also not guaranteed to see the most up-to-date value?

Reply:
Yeah perhaps, I thought about going full `ReadWriteLock` on the blob, but thought it might have been overkill.

Also now you mention it, this condition is redundant because the contents will never be null (I initialised contents to `new ByteArray(new byte[] {})` late in the piece.

Reply:
Removed the null check and made `contents` volatile in 64074243605fba90b06177a980da561089076a24